libindy-crypto/src/ffi/cl/verifier.rs

```diff
@@ -0,0 +1,349 @@
+use cl::verifier::*;
+use cl::*;
+use errors::ToErrorCode;
+use ffi::ErrorCode;
+use utils::ctypes::CTypesUtils;
+
+use libc::c_char;
+
+use std::os::raw::c_void;
+
+/// Creates random nonce
+///
+/// Note that nonce deallocation must be performed by calling indy_crypto_cl_nonce_free
+///
+/// # Arguments
+/// * `nonce_p` - Reference that will contain nonce instance pointer.
+#[no_mangle]
+pub extern fn indy_crypto_cl_verifier_new_nonce(nonce_p: *mut *const c_void) -> ErrorCode {
+    trace!("indy_crypto_cl_verifier_new_nonce: >>> {:?}", nonce_p);
+
+    check_useful_c_ptr!(nonce_p, ErrorCode::CommonInvalidParam1);
+
+    let res = match Verifier::new_nonce() {
+        Ok(nonce) => {
+            trace!("indy_crypto_cl_verifier_new_nonce: nonce: {:?}", nonce);
+            unsafe {
+                *nonce_p = Box::into_raw(Box::new(nonce)) as *const c_void;
+                trace!("indy_crypto_cl_verifier_new_nonce: *nonce_p: {:?}", *nonce_p);
+            }
+            ErrorCode::Success
+        }
+        Err(err) => err.to_error_code()
+    };
+
+    trace!("indy_crypto_cl_verifier_new_nonce: <<< res: {:?}", res);
+    res
+}
+
+
+/// Deallocates nonce instance.
+///
+/// # Arguments
+/// * `nonce_p` - Nonce instance pointer
+#[no_mangle]
+pub extern fn indy_crypto_cl_nonce_free(nonce: *const c_void) -> ErrorCode {
+    trace!("indy_crypto_cl_nonce_free: >>> nonce: {:?}", nonce);
+
+    check_useful_c_ptr!(nonce, ErrorCode::CommonInvalidParam1);
+
+    let nonce = unsafe { Box::from_raw(nonce as *mut Nonce); };
+    trace!("indy_crypto_cl_nonce_free: entity: nonce: {:?}", nonce);
+
+    let res = ErrorCode::Success;
+
+    trace!("indy_crypto_cl_nonce_free: <<< res: {:?}", res);
+    res
+}
+
+/// Creates and returns proof verifier.
+///
+/// Note that proof verifier deallocation must be performed by
+/// calling indy_crypto_cl_proof_verifier_finalize
+///
+/// # Arguments
+/// * `proof_verifier_p` - Reference that will contain proof verifier instance pointer.
+#[no_mangle]
+pub extern fn indy_crypto_cl_verifier_new_proof_verifier(proof_verifier_p: *mut *const c_void) -> ErrorCode {
+    trace!("indy_crypto_cl_verifier_new_proof_verifier: >>> {:?}", proof_verifier_p);
+
+    check_useful_c_ptr!(proof_verifier_p, ErrorCode::CommonInvalidParam1);
+
+    let res = match Verifier::new_proof_verifier() {
+        Ok(proof_verifier) => {
+            trace!("indy_crypto_cl_verifier_new_proof_verifier: proof_verifier: {:?}", proof_verifier);
+            unsafe {
+                *proof_verifier_p = Box::into_raw(Box::new(proof_verifier)) as *const c_void;
+                trace!("indy_crypto_cl_verifier_new_proof_verifier: *proof_verifier_p: {:?}", *proof_verifier_p);
+            }
+            ErrorCode::Success
+        }
+        Err(err) => err.to_error_code()
+    };
+
+    trace!("indy_crypto_cl_verifier_new_proof_verifier: <<< res: {:?}", res);
+    res
+}
+
+/// Add sub proof request to proof verifier.
```

Comment:
Please describe what sub-proof request is (this is part of proof request related to a schema-key)